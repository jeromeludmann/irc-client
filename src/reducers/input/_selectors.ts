import { createSelector } from "reselect";
import { selectInput } from "@app/reducers/channel/_selectors";

export const selectValue = createSelector(selectInput, input => input.value);
