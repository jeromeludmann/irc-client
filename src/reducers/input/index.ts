import {
  HistoryState,
  historyInitialState,
  reduceHistory,
} from "@app/reducers/input/history";
import {
  inputValueInitialState,
  reduceInputValue,
} from "@app/reducers/input/value";
import {
  inputDirtyValueInitialState,
  reduceInputDirtyValue,
} from "@app/reducers/input/dirtyValue";
import { Reducer, Action } from "redux";

export interface InputState {
  readonly value: string;
  readonly dirtyValue: string;
  readonly history: HistoryState;
}

export const inputInitialState: InputState = {
  value: inputValueInitialState,
  dirtyValue: inputDirtyValueInitialState,
  history: historyInitialState,
};

export const reduceInput: Reducer<InputState, Action> = (
  input = inputInitialState,
  action,
) => ({
  value: reduceInputValue(input.value, action, { input }),
  dirtyValue: reduceInputDirtyValue(input.dirtyValue, action, { input }),
  history: reduceHistory(input.history, action),
});
