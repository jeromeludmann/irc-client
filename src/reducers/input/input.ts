import reduceHistory, {
  HistoryState,
  HistoryAction,
  historyInitialState,
} from "@app/reducers/input/history";
import reduceValue, {
  ValueAction,
  valueInitialState,
} from "@app/reducers/input/value";
import reduceDirtyValue, {
  DirtyAction,
  dirtyInitialState,
} from "@app/reducers/input/dirty";

export interface InputState {
  readonly value: string;
  readonly dirtyValue: string;
  readonly history: HistoryState;
}

export type InputAction = ValueAction | DirtyAction | HistoryAction;

export const inputInitialState: InputState = {
  value: valueInitialState,
  dirtyValue: dirtyInitialState,
  history: historyInitialState,
};

export default (
  input = inputInitialState,
  action: InputAction,
): InputState => ({
  value: reduceValue(input.value, action as ValueAction, input),
  dirtyValue: reduceDirtyValue(
    input.dirtyValue,
    action as DirtyAction,
    input,
  ),
  history: reduceHistory(input.history, action as HistoryAction),
});
