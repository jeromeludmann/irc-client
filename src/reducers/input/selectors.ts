import { createSelector } from "reselect";
import { selectInput } from "@app/reducers/window/selectors";

export const selectValue = createSelector(selectInput, input => input.value);
