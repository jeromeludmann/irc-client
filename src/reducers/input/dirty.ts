import {
  UpdateInputValueAction,
  UPDATE_INPUT_VALUE,
  ENTER_INPUT_VALUE,
  EnterInputValueAction,
} from "@app/actions/input";
import { InputState } from "@app/reducers/input/input";
import { endOfHistory } from "@app/reducers/input/_helpers";

export type DirtyState = string;

export type DirtyAction = UpdateInputValueAction | EnterInputValueAction;

export const dirtyInitialState: DirtyState = "";

export default (
  dirty = dirtyInitialState,
  action: DirtyAction,
  input: InputState,
): DirtyState => {
  switch (action.type) {
    case UPDATE_INPUT_VALUE:
      return endOfHistory(input.history) ? action.payload.value : dirty;

    case ENTER_INPUT_VALUE:
      return "";

    default:
      return dirty;
  }
};
