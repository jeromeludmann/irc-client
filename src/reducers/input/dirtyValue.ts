import {
  UpdateInputValueAction,
  UPDATE_INPUT_VALUE,
  ENTER_INPUT_VALUE,
  EnterInputValueAction,
} from "@app/actions/ui";
import { InputState } from "@app/reducers/input";
import { endOfHistory } from "@app/reducers/input/_helpers";
import { Action } from "redux";

export type InputDirtyValueState = string;

export const inputDirtyValueInitialState: InputDirtyValueState = "";

type InputDirtyValueReducer<A = Action> = (
  dirtyValue: InputDirtyValueState,
  action: A,
  extraStates: { input: InputState },
) => InputDirtyValueState;

const updateInputValue: InputDirtyValueReducer<UpdateInputValueAction> = (
  dirtyValue,
  action,
  { input },
) => (endOfHistory(input.history) ? action.payload.value : dirtyValue);

const enterInputValue: InputDirtyValueReducer<EnterInputValueAction> = _ => "";

const map: { [action: string]: InputDirtyValueReducer } = {
  [UPDATE_INPUT_VALUE]: updateInputValue,
  [ENTER_INPUT_VALUE]: enterInputValue,
};

export const reduceInputDirtyValue = (
  dirtyValueState = inputDirtyValueInitialState,
  action: Action,
  extraStates: { input: InputState },
) =>
  map.hasOwnProperty(action.type)
    ? map[action.type](dirtyValueState, action, extraStates)
    : dirtyValueState;
