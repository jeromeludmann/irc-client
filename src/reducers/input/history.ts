import {
  GoBackInputHistoryAction,
  GoForwardInputHistoryAction,
  EnterInputValueAction,
  ENTER_INPUT_VALUE,
  GO_BACK_INPUT_HISTORY,
  GO_FORWARD_INPUT_HISTORY,
} from "@app/actions/ui";
import { beginOfHistory, endOfHistory } from "@app/reducers/input/_helpers";
import { mapReducer } from "@app/reducers/_map";
import { Action } from "redux";

export interface HistoryState {
  readonly values: string[];
  readonly index: number;
}

export const historyInitialState: HistoryState = {
  values: ["/connect -n localhost", "/connect localhost"], // TODO REMOVE ME
  index: 2,
};

type HistoryReducer<A = Action> = (
  input: HistoryState,
  action: A,
) => HistoryState;

const enterInputValue: HistoryReducer<EnterInputValueAction> = (
  history,
  action,
) => {
  const values = [...history.values, action.payload.value];
  return { values, index: values.length };
};

const goBackInputHistory: HistoryReducer<GoBackInputHistoryAction> = history =>
  beginOfHistory(history) ? history : { ...history, index: history.index - 1 };

const goForwardInputHistory: HistoryReducer<
  GoForwardInputHistoryAction
> = history =>
  endOfHistory(history) ? history : { ...history, index: history.index + 1 };

const map: { [action: string]: HistoryReducer } = {
  [ENTER_INPUT_VALUE]: enterInputValue,
  [GO_BACK_INPUT_HISTORY]: goBackInputHistory,
  [GO_FORWARD_INPUT_HISTORY]: goForwardInputHistory,
};

export const reduceHistory = mapReducer<HistoryState>(map);
