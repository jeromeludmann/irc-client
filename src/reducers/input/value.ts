import {
  ChangeInputValueAction,
  GoBackInputHistoryAction,
  GoForwardInputHistoryAction,
  CHANGE_INPUT_VALUE,
  GO_BACK_INPUT_HISTORY,
  GO_FORWARD_INPUT_HISTORY,
  SEND_INPUT_VALUE,
  SendInputValueAction,
} from "@app/actions/ui/input";
import { InputState } from "@app/reducers/input";
import { beginOfHistory, endOfHistory } from "@app/reducers/input/helpers";

export type ValueState = string;

export type ValueAction =
  | ChangeInputValueAction
  | SendInputValueAction
  | GoBackInputHistoryAction
  | GoForwardInputHistoryAction;

export const valueInitialState: ValueState = "";

export default (
  value = valueInitialState,
  action: ValueAction,
  input: InputState,
): ValueState => {
  switch (action.type) {
    case CHANGE_INPUT_VALUE:
      return action.payload.value;

    case SEND_INPUT_VALUE:
      return "";

    case GO_BACK_INPUT_HISTORY:
      return beginOfHistory(input.history)
        ? value
        : input.history.values[input.history.index - 1];

    case GO_FORWARD_INPUT_HISTORY:
      return endOfHistory(input.history)
        ? value
        : input.history.values[input.history.index + 1] || input.dirtyValue;

    default:
      return value;
  }
};
