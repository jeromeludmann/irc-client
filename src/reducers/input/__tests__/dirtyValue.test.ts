import { reduceInputDirtyValue } from "@app/reducers/input/dirtyValue";
import { updateInputValue, enterInputValue } from "@app/actions/ui";

describe("reduce input dirty value by dispatching", () => {
  const extraStates = {
    input: {
      value: "",
      dirtyValue: "",
      history: { index: 0, values: [] },
    },
  };

  describe("update input value", () => {
    test("anywhere in the history", () => {
      expect(
        reduceInputDirtyValue(undefined, updateInputValue("hello"), {
          input: {
            value: "",
            dirtyValue: "",
            history: { index: 1, values: ["hello", "world"] },
          },
        }),
      ).toMatchSnapshot();
    });

    test("at the end of history", () => {
      expect(
        reduceInputDirtyValue(undefined, updateInputValue("hello"), {
          input: {
            value: "",
            dirtyValue: "",
            history: { index: 2, values: ["hello", "world"] },
          },
        }),
      ).toMatchSnapshot();
    });
  });

  test("enter input value", () => {
    expect(
      reduceInputDirtyValue(undefined, enterInputValue("hello"), extraStates),
    ).toMatchSnapshot();
  });
});
