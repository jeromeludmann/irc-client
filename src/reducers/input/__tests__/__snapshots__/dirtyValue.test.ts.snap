// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reduce input dirty value by dispatching enter input value 1`] = `""`;

exports[`reduce input dirty value by dispatching update input value anywhere in the history 1`] = `""`;

exports[`reduce input dirty value by dispatching update input value at the end of history 1`] = `"hello"`;
